File: csharp/demo/demo22a-blocking-queue.cs
/*
 * BLOCKING QUEUES
 * Version A: A slow producer and a fast consumer
 */
using System;
using System.Collections.Concurrent;
using System.Threading;



class Demo22A : IRunnable
{
    public void run()
    {
        // blocking queue with capacity = 2
        var queue = new BlockingCollection<string>(2);

        new Thread(() => producer(queue)).Start();
        new Thread(() => consumer(queue)).Start();

        // Should call queue.Dispose();
    }


    private void producer(BlockingCollection<string> queue)
    {
        Thread.Sleep(2000);
        queue.Add("Alice");

        Thread.Sleep(2000);
        queue.Add("likes");

        Thread.Sleep(2000);
        queue.Add("singing");
    }


    private void consumer(BlockingCollection<string> queue)
    {
        string data;

        for (int i = 0; i < 3; ++i)
        {
            Console.WriteLine("\nWaiting for data...");

            data = queue.Take();

            Console.WriteLine("    " + data);
        }
    }
}

File: python/demo22a_blocking_queue.py
'''
BLOCKING QUEUES
Version A: A slow producer and a fast consumer
'''

import time
from queue import Queue
import threading



def producer(q: Queue):
    time.sleep(2)
    q.put('Alice')

    time.sleep(2)
    q.put('likes')

    time.sleep(2)
    q.put('singing')



def consumer(q: Queue):
    for _ in range(3):
        print('\nWaiting for data...')
        data = q.get()
        print(f'    {data}')



blkq = Queue()

threading.Thread(target=producer, args=(blkq,)).start()
threading.Thread(target=consumer, args=(blkq,)).start()
